File: algorithms/dynamic-programming/maximum-subarray.py
class Solution:
  def maxSubArray(self, nums: List[int]) -> int:
    length = len(nums)
    cache = [0] * length

    cache[0] = nums[0]
    maxValue = nums[0]
    i = 1
    for i in range(1, length):
      cache[i] = max(cache[i-1] + nums[i], nums[i])

      if maxValue < cache[i]:
        maxValue = cache[i]

    return maxValue

File: 30-day-leetcode-challenge/april-2020/maximum-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function(nums) {
  let length = nums.length;
  let maxAtPosition = new Array(length);
  let max = nums[0];

  maxAtPosition[0] = nums[0];
  for (let i = 1; i < length; i++) {
    maxAtPosition[i] = Math.max(maxAtPosition[i-1] + nums[i], nums[i] );
    max = Math.max(max, maxAtPosition[i]);
  }

  return max;
};
